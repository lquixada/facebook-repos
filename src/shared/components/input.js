import React from 'react'
import styled from 'styled-components'

export const Input = (props) => (
  <TextInput type='text' {...props} />
)

const TextInput = styled.input`
  background-color: #33373c;
  border: .1rem solid;
  border-radius: .5rem;
  border-color: #303030 #505050 #505050 #303030;
  width: 30rem;
  padding: 0rem 1rem .5rem;
  color: #b9b9b9;
  font-family: 'Lato', Helvetica, sans-serif;
  font-size: 3rem;
  letter-spacing: -.1rem;
  border-radius: .4rem;
  outline: none;

  @media (max-width: 576px) {
    margin: 1rem 0 0;
    width: 100%;
  }
`
