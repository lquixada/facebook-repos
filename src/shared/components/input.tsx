import styled from 'styled-components'

export const Input = styled.input.attrs({
  'aria-label': 'Repository input',
  'autoFocus': true,
  'type': 'text',
})`
  background-color: #33373c;
  border: 1px solid;
  border-radius: 5px;
  border-color: #303030 #505050 #505050 #303030;
  width: 300px;
  padding: 0 10px 5px;
  color: #b9b9b9;
  font-family: 'Lato', Helvetica, sans-serif;
  font-size: 30px;
  letter-spacing: -1px;
  border-radius: 4px;
  outline: none;

  @media (max-width: 576px) {
    margin: 10px 0 0;
    width: 100%;
  }
`
