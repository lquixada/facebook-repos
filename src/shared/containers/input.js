import React from 'react'
import {withRouter} from 'react-router'

import {Input} from '../components/input'

export class InputContainer extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      value: '',
      minLength: 2
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
  }

  // Renders current owner on server-side rendering
  componentWillMount () {
    this.setCurrentOwner()
  }

  // Renders current owner on browser navigation
  componentDidUpdate (prevProps) {
    if (this.hasOwnerChanged(prevProps)) {
      this.setCurrentOwner()
    }
  }

  setCurrentOwner () {
    this.setState({
      value: this.getOwner(this.props.location)
    })
  }

  getOwner (location) {
    const match = /\/r\/(\w+)\/?/.exec(location.pathname)

    if (match) {
      return match[1]
    }
  }

  hasOwnerChanged (prevProps) {
    return this.getOwner(prevProps.location) !== this.getOwner(this.props.location)
  }

  handleChange (evt) {
    const value = evt.target ? evt.target.value : evt
    this.setState({value: value.trim()})
  }

  handleKeyUp (evt) {
    if (this.state.value.length < this.state.minLength) {
      return
    }

    if (evt.which === 13) {
      this.navigate()
    }
  }

  navigate () {
    this.props.history.push(`/r/${this.state.value}`)
  }

  render () {
    return (
      <Input value={this.state.value}
        onChange={this.handleChange}
        onKeyUp={this.handleKeyUp} />
    )
  }
}

export default withRouter(InputContainer)
