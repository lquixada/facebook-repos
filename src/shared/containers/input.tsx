import * as H from 'history'
import React from 'react'
import {withRouter} from 'react-router'

import {Input} from '../components/input'

interface IState {
  minLength: number
  value: string
}

export class InputContainer extends React.Component<any, IState, any> {
  constructor(props) {
    super(props)
    this.state = {
      minLength: 2,
      value: '',
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleKeyUp = this.handleKeyUp.bind(this)
  }

  // Renders current owner on server-side rendering
  public componentWillMount() {
    this.setCurrentOwner()
  }

  // Renders current owner on browser navigation
  public componentDidUpdate(prevProps) {
    if (this.hasOwnerChanged(prevProps)) {
      this.setCurrentOwner()
    }
  }

  public setCurrentOwner() {
    this.setState({
      value: this.getOwner(this.props.location),
    })
  }

  public getOwner(location) {
    const match = /\/r\/(\w+)\/?/.exec(location.pathname)

    if (match) {
      return match[1]
    }

    return ''
  }

  public hasOwnerChanged(prevProps) {
    return this.getOwner(prevProps.location) !== this.getOwner(this.props.location)
  }

  public handleChange(evt) {
    const value = evt.target ? evt.target.value : evt
    this.setState({value: value.trim()})
  }

  public handleKeyUp(evt) {
    if (this.state.value.length < this.state.minLength) {
      return
    }

    if (evt.which === 13) {
      this.navigate()
    }
  }

  public navigate() {
    this.props.history.push(`/r/${this.state.value}`)
  }

  public render() {
    return (
      <Input value={this.state.value}
        onChange={this.handleChange}
        onKeyUp={this.handleKeyUp} />
    )
  }
}

export default withRouter(InputContainer)
