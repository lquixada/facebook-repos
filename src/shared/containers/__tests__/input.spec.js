import React from 'react'
import {mount} from 'enzyme'
import {createLocation} from 'history'

import {InputContainer} from '../input'

describe('<InputContainer />', () => {
  let component
  let history
  let location

  beforeEach(() => {
    location = createLocation('/r/owner1')
    history = {
      push: jest.fn()
    }
    component = mount(<InputContainer location={location} history={history} />)
  })

  it('renders input with owner name', () => {
    expect(component).toMatchSnapshot()
  })

  it('renders new owner name when location changes', () => {
    component.setProps({
      location: createLocation('/r/owner2')
    })
    expect(component.update()).toMatchSnapshot()
  })

  it('navigate when user presses Enter', () => {
    component.find('input').simulate('change', {target: {value: 'some-owner'}})
    component.find('input').simulate('keyup', {which: 13}) // Enter

    expect(history.push).toBeCalledWith('/r/some-owner')
  })

  it('does not navigate when user types owner smaller than 2 characters', () => {
    component.find('input').simulate('change', {target: {value: 'o'}})
    component.find('input').simulate('keyup', {which: 13}) // Enter

    expect(history.push).not.toBeCalled()
  })
})
