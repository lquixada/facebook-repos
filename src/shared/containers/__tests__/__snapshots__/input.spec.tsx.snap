// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer /> renders input with owner name 1`] = `
<InputContainer
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/r/owner1",
      "search": "",
      "state": undefined,
    }
  }
>
  <styled.input
    onChange={[Function]}
    onKeyUp={[Function]}
    value="owner1"
  >
    <input
      aria-label="Repository input"
      autoFocus={true}
      className="sc-bdVaJa cdFJtR"
      onChange={[Function]}
      onKeyUp={[Function]}
      type="text"
      value="owner1"
    />
  </styled.input>
</InputContainer>
`;

exports[`<InputContainer /> renders new owner name when location changes 1`] = `
<InputContainer
  history={
    Object {
      "push": [MockFunction],
    }
  }
  location={
    Object {
      "hash": "",
      "pathname": "/r/owner2",
      "search": "",
      "state": undefined,
    }
  }
>
  <styled.input
    onChange={[Function]}
    onKeyUp={[Function]}
    value="owner2"
  >
    <input
      aria-label="Repository input"
      autoFocus={true}
      className="sc-bdVaJa cdFJtR"
      onChange={[Function]}
      onKeyUp={[Function]}
      type="text"
      value="owner2"
    />
  </styled.input>
</InputContainer>
`;
